Frameworks/CoreGraphics/CGDataProvider.mm

```diff
@@ -16,130 +16,395 @@
+#import <vector>
+
+// Determines what amount of bytes to try and read at a time for indirect access data providers
+static constexpr size_t c_streamSize = 2048;
+
+#pragma region __CGDataProviderInternal
+namespace __CGDataProviderInternal {
+class __CGDataProviderInternal {
+    mutable std::unique_ptr<CFHashCode> m_hashValue;
+
+public:
+    virtual ~__CGDataProviderInternal() {
     }
-    filename = nil;
-    [super dealloc];
-}
+    virtual const void* GetData() const = 0;
+    virtual size_t GetSize() const = 0;
+    bool operator==(const __CGDataProviderInternal& rhs) const {
+        if (GetSize() != rhs.GetSize()) {
+            return false;
+        }
 
-- (NSString*)_fileName {
-    return filename;
-}
+        return memcmp(GetData(), rhs.GetData(), GetSize()) == 0;
+    }
 
-- (instancetype)init {
-    return [self initWithBytes:"" length:0];
-}
+    CFHashCode Hash() const {
+        if (!m_hashValue) {
+            m_hashValue.reset(new CFHashCode(CFHashBytes((UInt8*)GetData(), GetSize())));
+        }
 
-- (instancetype)initWithBytesNoCopy:(void*)bytes length:(NSUInteger)length freeWhenDone:(BOOL)freeWhenDone {
-    if (self = [super init]) {
-        _data.attach([[NSData alloc] initWithBytesNoCopy:bytes length:length freeWhenDone:freeWhenDone]);
+        return *m_hashValue;
     }
-    return self;
-}
+};
 
-- (const void*)bytes {
-    return [_data bytes];
-}
+class __UserDataProvider : public __CGDataProviderInternal {
+    mutable void* m_info;
+    const CGDataProviderCallbacks* m_callbacks;
+    mutable std::vector<unsigned char> m_data;
 
-- (NSUInteger)length {
-    return [_data length];
-}
+public:
+    __UserDataProvider(void* info, const CGDataProviderCallbacks* callbacks) : m_info(info), m_callbacks(callbacks) {
+    }
 
-@end
+    const void* GetData() const override {
+        if (m_data.empty()) {
+            size_t count = 0, delta = 0;
+            do {
+                m_data.resize(m_data.size() + c_streamSize);
+                delta = m_callbacks->getBytes(m_info, m_data.data() + count, c_streamSize);
+                count += delta;
+                m_callbacks->skipBytes(m_info, delta);
+            } while (delta == c_streamSize);
+            m_data.resize(count);
+        }
 
-/**
- @Status Caveat
- @Notes Only supports file:/// URLs
-*/
-CGDataProviderRef CGDataProviderCreateWithURL(CFURLRef url) {
-    NSString* path = [static_cast<NSURL*>(url) path];
-    CGDataProvider* ret = [[CGDataProvider alloc] initWithContentsOfFile:path];
-    if (ret) {
-        ret->filename = path;
+        return m_data.data();
     }
 
-    return ret;
-}
+    size_t GetSize() const override {
+        if (m_data.empty()) {
+            GetData();
+        }
+
+        return m_data.size();
+    }
+
+    ~__UserDataProvider() {
+        m_callbacks->releaseProvider(m_info);
+    }
+};
+
+class __RawDataProvider : public __CGDataProviderInternal {
+    void* m_info;
+    const void* m_data;
+    size_t m_size;
```

Comment:
do we have to take a copy?

---
In reply to: [92906947](https://github.com/Microsoft/WinObjC/pull/1579#discussion_r92906947) [](ancestors = 92906947)